runtime/src/iree/builtins/mmt4d/mmt4d.h

```diff
@@ -0,0 +1,148 @@
+//
+// Avoid #ifdef entirely where possible: they indicate a leakage of host build
+// configuration into what is supposed to be a portable module. Anything that
+// requires target-specific conditional logic must be implemented via an extern
+// that can be substituted by the IREE compiler when producing the final
+// target-specific module.
+
+// We require that this header compile on bare-metal targets with no stdlib.
+// These two headers are clean and do not include any other headers:
+#include "iree/base/attributes.h"
+#include "iree/base/target_platform.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// Target architecture selection
+//===----------------------------------------------------------------------===//
+// The "generic" target is used if no other target is specified. All platforms
+// should support the generic path and then optionally provide their own
+// specializations as needed. The generic path can be forced by passing
+// -DIREE_MMT4D_ARCH_GENERIC_32=1 or -DIREE_MMT4D_ARCH_GENERIC_64=1 to the
+// compiler in addition to -DIREE_PLATFORM_GENERIC=1.
+
+#if defined(IREE_MMT4D_ARCH_GENERIC_32)
+#define IREE_MMT4D_SIZE_TYPE int32_t
+#elif defined(IREE_MMT4D_ARCH_GENERIC_64)
+#define IREE_MMT4D_SIZE_TYPE int64_t
+#elif defined(IREE_ARCH_ARM_64)
+#define IREE_MMT4D_ARCH_ARM_64 1
+#define IREE_MMT4D_SIZE_TYPE int64_t
+#else
+#define IREE_MMT4D_ARCH_GENERIC_64 1
+#define IREE_MMT4D_SIZE_TYPE int64_t
+#endif  // IREE_ARCH_*
+
+//===----------------------------------------------------------------------===//
+// Attributes and metadata
+//===----------------------------------------------------------------------===//
+
+// Tagged on functions that are part of the public API.
+// TODO(benvanik): use this to change symbol visibility? We don't want a library
+// that embeds this one to transitively export functions but may want to have
+// the functions exported based on how we link them. For now this is used as
+// documentation.
+#define IREE_MMT4D_EXPORT
+
+//===----------------------------------------------------------------------===//
+// stdint.h
+//===----------------------------------------------------------------------===//
+// https://pubs.opengroup.org/onlinepubs/009604599/basedefs/stdint.h.html
+// NOTE: prefer not using size_t/ptrdiff_t/etc (as they are target dependent).
+// We avoid including the toolchain file as it may not match with the target
+// information we have in the IREE compiler. They're also generally tire fires
+// that significantly bloat the bitcode we build and link into binaries: stdint
+// on Windows includes windows.h, for example.
+//
+// NOTE: callers must #include <stdint.h> before this header; unfortunately
+// there's not a great way to redefine these in the absence of stdint.h that
+// also operates with stdint.h.
+
+#if !defined(INT8_MIN)
+
+typedef signed char int8_t;
```

Comment:
So even `<stdint.h>` is off-limits here? And users of this header will never themselves include `<stdint.h>` (and potentially get conflicting typedefs?)

Reply:
Today yes - stdint pulls in tons of toolchain stuff we want to avoid when going to bitcode as otherwise we fixate the bitcode to the compiler and toolchain that produced the bitcode instead of the one we use in iree-compile and things go bad fast. The goal is that the core of the library can be compiled once per major architecture instead of once per compiler x stdlib x toolchain x host platform x compilation modes etc.

We can selectively lighten this restriction in the future but there's not really much worth pulling in from the local toolchain when it comes to microkernels - the tests can still include whatever they want/use C++/etc.